packages/output/src/browser/output-resource.ts

```diff
@@ -0,0 +1,58 @@
+/********************************************************************************
+ * Copyright (C) 2020 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import URI from '@theia/core/lib/common/uri';
+import { Event, Resource, ResourceReadOptions, DisposableCollection, Emitter } from '@theia/core/lib/common';
+import { Deferred } from '@theia/core/lib/common/promise-util';
+import { MonacoEditorModel } from '@theia/monaco/lib/browser/monaco-editor-model';
+
+export class OutputResource implements Resource {
+
+    protected _textModel: monaco.editor.ITextModel | undefined;
+    protected onDidChangeContentsEmitter = new Emitter<void>();
+    protected toDispose = new DisposableCollection(
+        this.onDidChangeContentsEmitter
+    );
+
+    constructor(readonly uri: URI, readonly editorModel: Deferred<MonacoEditorModel>) {
```

Comment:
we should better to pass `MonacoEditorModel` here, otherwise there can be a race that resource disposed but model was not resolved yet.

Reply:
> we should better pass MonacoEditorModel here

I am with you but we can't, that's the point. If you have a better idea without changing the API too much, we can adjust it. But such vague advice "we should not pass" does not help much.

From [here](https://github.com/eclipse-theia/theia/pull/7570/files#diff-c288f39cebca6c7afc5167fe59b166ebR194-R196):
> // We have to register the resource first, because `textModelService#createModelReference` will require it
        // right after creating the monaco.editor.ITextModel.
        // All `append` and `appendLine` will be deferred until the underlying text-model instantiation.

Reply:
ok, please add a check in `then` closure that if `toDispose.disposed` already then we should dispose the model and exit.

Thinking about it we should not dispose a model but a reference. Model will be auto disposed when all references are disposed. If we do differently there can be another race condition there a channel can end up with a disposed model.

Reply:
Let's do that. Thanks!

Reply:
> Thinking about it we should not dispose a model but a reference.

Great pointer. Well, it was a bug. I have fixed it now.  👍 We cannot pass the deferred `MonacoEditorModel`, but I had to use the `monaco.editor.IReference<MonacoEditorModel>` instead.

> please add a check in `then` closure that if `toDispose.disposed`

Sorry, which `toDispose.disposed` do you mean?

Reply:
I meant:
```ts
this.editorModelRef.promise.then(modelRef => {
            if (this.toDispose.disposed) {
                modelRef.dispose();            
                return;
            }
            const textModel = modelRef.object.textEditorModel;
            this._textModel = textModel;
            this.toDispose.push(modelRef);
            this.toDispose.push(this._textModel!.onDidChangeContent(() => this.onDidChangeContentsEmitter.fire()));
        });
```